binding.cc

```diff
@@ -2,189 +2,147 @@
-#include "extensions/tweak/tweak.h"
 #include "extensions/pbkdf2/pbkdf2.h"
+#include "extensions/tweak/tweak.h"
 #include "sodium/crypto_generichash.h"
 
-static uint8_t typedarray_width (js_typedarray_type_t type) {
-  switch (type) {
-    case js_int8array: return 1;
-    case js_uint8array: return 1;
-    case js_uint8clampedarray: return 1;
-    case js_int16array: return 2;
-    case js_uint16array: return 2;
-    case js_int32array: return 4;
-    case js_uint32array: return 4;
-    case js_float32array: return 4;
-    case js_float64array: return 8;
-    case js_bigint64array: return 8;
-    case js_biguint64array: return 8;
-    default: return 0;
-  }
-}
-
-js_value_t *
-sn_sodium_memzero (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, sodium_memzero)
-
-  SN_ARGV_TYPEDARRAY(buf, 0)
-
-  sodium_memzero(buf_data, buf_size);
+#define assert_bounds(arraybuffer) \
+  assert(arraybuffer##_offset + arraybuffer##_len <= arraybuffer.size())
 
-  return NULL;
+static inline void
+sn_sodium_memzero(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> buf
+) {
+  sodium_memzero(buf.data(), buf.size_bytes());
 }
 
-js_value_t *
-sn_sodium_mlock (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, sodium_mlock)
-
-  SN_ARGV_TYPEDARRAY(buf, 0)
-
-  SN_RETURN(sodium_mlock(buf_data, buf_size), "memory lock failed")
+static inline int
+sn_sodium_mlock(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> buf
+) {
+  return sodium_mlock(buf.data(), buf.size_bytes());
 }
 
-js_value_t *
-sn_sodium_munlock (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, sodium_munlock)
-
-  SN_ARGV_TYPEDARRAY(buf, 0)
-
-  SN_RETURN(sodium_munlock(buf_data, buf_size), "memory unlock failed")
+static inline int
+sn_sodium_munlock(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> buf
+) {
+  return sodium_munlock(buf.data(), buf.size_bytes());
 }
 
-static void sn_sodium_free_finalise (js_env_t *env, void *finalise_data, void *finalise_hint) {
+static void
+sn_external_arraybuffer_finalize(js_env_t *env, void *finalise_data, void *finalise_hint) {
   sodium_free(finalise_data);
-
-  int64_t ext_mem;
-  int err = js_adjust_external_memory(env, -4 * 4096, &ext_mem);
-  assert(err == 0);
 }
 
-js_value_t *
-sn_sodium_free (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, sodium_free)
-
-  SN_ARGV_TYPEDARRAY_PTR(buf, 0)
-  if (buf_data == NULL) return NULL;
-
-  js_value_t *array_buf;
-  SN_STATUS_THROWS(js_get_named_property(env, argv[0], "buffer", &array_buf), "failed to get arraybuffer");
+static js_arraybuffer_t
+sn_sodium_malloc(js_env_t *env, js_receiver_t, uint32_t len) {
+  void *ptr = sodium_malloc(len);
+  assert(ptr != nullptr);
 
-  SN_STATUS_THROWS(js_detach_arraybuffer(env, array_buf), "failed to detach array buffer");
-
-  void *ptr;
-  err = js_remove_wrap(env, array_buf, &ptr);
+  js_value_t *buffer;
+  int err = js_create_external_arraybuffer(env, ptr, len, sn_external_arraybuffer_finalize, NULL, &buffer);
```

Comment:
There's a C++ version of this: https://github.com/holepunchto/libjstl/blob/main/test/create-external-arraybuffer-with-finalizer.cc